jetty-server/src/main/java/org/eclipse/jetty/server/DetectorConnectionFactory.java

```diff
@@ -0,0 +1,294 @@
+    {
+        return configure(new DetectorConnection(endPoint, connector), connector, endPoint);
+    }
+
+    private class DetectorConnection extends AbstractConnection implements Connection.UpgradeFrom, Connection.UpgradeTo
+    {
+        private final Connector _connector;
+        private final ByteBuffer _buffer;
+
+        private DetectorConnection(EndPoint endp, Connector connector)
+        {
+            super(endp, connector.getExecutor());
+            _connector = connector;
+            _buffer = connector.getByteBufferPool().acquire(getInputBufferSize(), true);
+        }
+
+        @Override
+        public void onUpgradeTo(ByteBuffer prefilled)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Detector {} copying prefilled buffer {}", getProtocol(), BufferUtil.toDetailString(prefilled));
+            if (BufferUtil.hasContent(prefilled))
+                BufferUtil.append(_buffer, prefilled);
+        }
+
+        @Override
+        public ByteBuffer onUpgradeFrom()
+        {
+            return _buffer;
+        }
+
+        @Override
+        public void onOpen()
+        {
+            super.onOpen();
+            if (!detectAndUpgrade())
+                fillInterested();
+        }
+
+        @Override
+        public void onFillable()
+        {
+            try
+            {
+                while (BufferUtil.space(_buffer) > 0)
+                {
+                    // Read data
+                    int fill = getEndPoint().fill(_buffer);
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Detector {} filled buffer with {} bytes", getProtocol(), fill);
+                    if (fill < 0)
+                    {
+                        _connector.getByteBufferPool().release(_buffer);
+                        getEndPoint().shutdownOutput();
+                        return;
+                    }
+                    if (fill == 0)
+                    {
+                        fillInterested();
+                        return;
+                    }
+
+                    if (detectAndUpgrade())
+                        return;
+                }
+
+                // all Detecting instances want more bytes than this buffer can store
+                LOG.warn("Detector {} failed to detect upgrade target on {} for {}", getProtocol(), _detectingConnectionFactories, getEndPoint());
+                releaseAndClose();
+            }
+            catch (Throwable x)
+            {
+                LOG.warn("Detector {} error for {}", getProtocol(), getEndPoint(), x);
+                releaseAndClose();
+            }
+        }
+
+        /**
+         * @return true when upgrade was performed, false otherwise.
+         */
+        private boolean detectAndUpgrade()
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Detector {} performing detection with {} bytes", getProtocol(), _buffer.remaining());
```

Comment:
Please add an early return testing for `BufferUtil.isEmpty(_buffer)`.

Reply:
Done.